src/main/java/it/gov/pagopa/reward/service/reward/RuleEngineServiceImpl.java

```diff
@@ -26,52 +30,65 @@ public class RuleEngineServiceImpl implements RuleEngineService {
                                  RewardContextHolderService rewardContextHolderService,
                                  Transaction2TransactionDroolsMapper transaction2TransactionDroolsMapper,
-                                 TransactionDroolsDTO2RewardTransactionMapper transactionDroolsDTO2RewardTransactionMapper) {
+                                 TransactionDroolsDTO2RewardTransactionMapper transactionDroolsDTO2RewardTransactionMapper, TransactionProcessedRepository transactionProcessedRepository) {
         this.ruleEngineConfig = ruleEngineConfig;
         this.rewardContextHolderService = rewardContextHolderService;
         this.transaction2TransactionDroolsMapper = transaction2TransactionDroolsMapper;
         this.transactionDroolsDTO2RewardTransactionMapper = transactionDroolsDTO2RewardTransactionMapper;
+        this.transactionProcessedRepository = transactionProcessedRepository;
     }
 
     @Override
     public RewardTransactionDTO applyRules(TransactionDTO transaction, List<String> initiatives, UserInitiativeCounters userInitiativeCounters) {
+
         TransactionDroolsDTO trx = transaction2TransactionDroolsMapper.apply(transaction);
 
-        if(!initiatives.isEmpty()){
-            StatelessKieSession statelessKieSession = rewardContextHolderService.getRewardRulesKieContainer().newStatelessKieSession();
-
-            trx.setInitiatives(initiatives);
-            trx.setRewards(new HashMap<>());
-
-            List<Command<?>> cmds = new ArrayList<>();
-            cmds.add(CommandFactory.newInsert(ruleEngineConfig));
-            cmds.add(CommandFactory.newInsert(userInitiativeCounters));
-            cmds.add(CommandFactory.newInsert(trx));
-            for (String initiative: initiatives) {
-                cmds.add(new AgendaGroupSetFocusCommand(initiative));
-            }
-
-            long before=System.currentTimeMillis();
-            statelessKieSession.execute(CommandFactory.newBatchExecution(cmds));
-            long after=System.currentTimeMillis();
-
-            log.info("[PERFORMANCE_LOG] - Time between before and after fireAllRules %d ms with transactionId: {idTrxAcquirer: %s, acquirerCode: %s, trxDate: %s}".formatted(after-before, transaction.getIdTrxAcquirer(),transaction.getAcquirerCode(),transaction.getTrxDate()));
-
-            /* TODO uncomment
-            if(log.isDebugEnabled()){
-                log.debug("Time between before and after fireAllRules: {} ms", after-before);
-            }
-            */
-
-            log.debug("Send message prepared: {}", trx);
-            log.info("Transaction evaluated  ({}) and resulted into rewards:({}), initiativeRejectionReason:{}", "%s-%s".formatted(trx.getHpan(), trx.getTrxDate()), trx.getRewards(), trx.getInitiativeRejectionReasons());
-        }else {
-            trx.getRejectionReasons().add(RewardConstants.TRX_REJECTION_REASON_NO_INITIATIVE);
-        }
+        transactionProcessedRepository.findById(transaction.getIdTrxAcquirer())
```

Comment:
these way you are evaluating the transaction only if it is duplicate!
you should log an ERROR describing the duplicate condition in case of flatMap, while you should continue the execution Only if empty you should continue the elaboration